Update MSAL libraries and use new MSALRuntime-based broker implementation

mjcheetham opened the pull request:
Use the new Windows broker which is based on the MSALRuntime; an export wrapper around a native, cross-platform MSAL library.

In this new set up, we drop the `.Desktop` package in favour of the `.Broker` package that also means we drop the WebView2Loader.dll, which we didn't make use of anyway. There are a few new binaries to be distributed in the new model, including a P/Invoke layer, IdentityModel abstractions library, and the native `msalruntime_x86.dll`.

Note that GCM still only support x86 on Windows, and only supports broker use on Windows. For this reason we don't bother adding the broker package on non-.NET Framework builds to keep the sizes on Mac/Linux to a minimum.

Also update the MSAL extensions library whilst we are here to pick up various bug fixes, and drop workarounds that are no longer required for this new broker implementation!

mjcheetham committed:
msauth: update MSAL library packages to latest

mjcheetham committed:
msauth: use new Windows broker based on MSALRuntime

Use the new Windows broker which is based on the MSALRuntime; an export
wrapper around a native, cross-platform MSAL library.

In this new set up, we drop the `.Desktop` package in favour of the
`.Broker` package that also means we drop the WebView2Loader.dll, which
we didn't make use of anyway. There are a few new binaries to be
distrubuted in the new model, including a P/Invoke layer, IdentityModel
abstractions library, and the native msalruntime_x86.dll.

Note that GCM still only support x86 on Windows, and only supports
broker use on Windows. For this reason we don't bother adding the broker
package on non-.NET Framework builds to keep the sizes on Mac/Linux to a
minimium.

mjcheetham committed:
msauth: always set the parent window for MSAL

Always set the parent window handle for MSAL on Windows.

Previously we only set this if provided a handle by the user/config. Now
however we must always try and provide a handle because using the new
MSALRuntime-based Windows broker means we must do so - MSAL no longer
provides us with a 'dummy' handle to use.

Use the parent console window handle, as recommended by the MSAL docs:
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/wam#parent-window-handles

mjcheetham commented:
@bgavrilMS could a member of the MSAL team have a quick glance over our usage of the new broker implementation on Windows, to make sure we're not doing something wrong? Thanks!

bgavrilMS commented on src/shared/Core/Authentication/MicrosoftAuthentication.cs:
```diff
@@ -41,55 +42,6 @@ public class MicrosoftAuthentication : AuthenticationBase, IMicrosoftAuthenticat
             "live", "liveconnect", "liveid",
         };
 
-        #region Broker Initialization
```
deleting workounds, so satisfying!

bgavrilMS commented on src/shared/Core/Authentication/MicrosoftAuthentication.cs:
```diff
@@ -279,20 +224,40 @@ private async Task<AuthenticationResult> GetAccessTokenSilentlyAsync(IPublicClie
+                if (!string.IsNullOrWhiteSpace(Context.Settings.ParentWindowId) &&
+                    int.TryParse(Context.Settings.ParentWindowId, out int hWndInt) && hWndInt > 0)
+                {
+                    parentHandle = new IntPtr(hWndInt);
```
How are you calculating the `hwndInt` ? For console apps it's a bit trickier - see https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/wam#parent-window-handles

bgavrilMS commented on src/shared/Core/Authentication/MicrosoftAuthentication.cs:
```diff
@@ -279,20 +224,40 @@ private async Task<AuthenticationResult> GetAccessTokenSilentlyAsync(IPublicClie
+                else
+                {
+                    IntPtr consoleHandle = Kernel32.GetConsoleWindow();
+                    parentHandle = User32.GetAncestor(consoleHandle, GetAncestorFlags.GetRootOwner);
```
+1

ldennington commented on src/shared/Core/Core.csproj:
```diff
@@ -13,13 +13,13 @@
     <Reference Include="System.Net.Http" />
     <Reference Include="System.Web" />
-    <PackageReference Include="Microsoft.Identity.Client.Desktop" Version="4.37.0" />
+    <PackageReference Include="Microsoft.Identity.Client.Desktop" Version="4.52.0" />
```
This means one less thing to worry about with #1143. Thanks!

ldennington commented on src/windows/Installer.Windows/Setup.iss:
```diff
@@ -125,20 +125,19 @@ Source: "{#PayloadDir}\Microsoft.AzureRepos.dll";                       DestDir:
 Source: "{#PayloadDir}\gcmcoreui.dll";                                  DestDir: "{app}"; Flags: ignoreversion
 Source: "{#PayloadDir}\gcmcoreuiwpf.dll";                               DestDir: "{app}"; Flags: ignoreversion
-Source: "{#PayloadDir}\Microsoft.Identity.Client.Desktop.dll";          DestDir: "{app}"; Flags: ignoreversion
+Source: "{#PayloadDir}\Microsoft.Identity.Client.Broker.dll";           DestDir: "{app}"; Flags: ignoreversion
```
Nit: Is the extra space at the end on purpose?

mjcheetham replied:
Yes, to keep the alignment of `DestDir: ...`.

ldennington commented on src/windows/Installer.Windows/Setup.iss:
```diff
@@ -125,20 +125,19 @@ Source: "{#PayloadDir}\Microsoft.AzureRepos.dll";                       DestDir:
-Source: "{#PayloadDir}\Microsoft.Web.WebView2.Wpf.dll";                 DestDir: "{app}"; Flags: ignoreversion
+Source: "{#PayloadDir}\Microsoft.Identity.Client.NativeInterop.dll";    DestDir: "{app}"; Flags: ignoreversion
+Source: "{#PayloadDir}\Microsoft.IdentityModel.Abstractions.dll";       DestDir: "{app}"; Flags: ignoreversion
+Source: "{#PayloadDir}\msalruntime_x86.dll";                            DestDir: "{app}"; Flags: ignoreversion
```
Nit: extra spaces

mjcheetham replied:
Needed to keep alignment.

ldennington replied:
Ah I see - thanks for clarifying.

ldennington commented on src/shared/Core/Authentication/MicrosoftAuthentication.cs:
```diff
@@ -223,12 +224,25 @@ private async Task<AuthenticationResult> GetAccessTokenSilentlyAsync(IPublicClie
+                }
+                else
+                {
+                    IntPtr consoleHandle = Kernel32.GetConsoleWindow();
+                    parentHandle = User32.GetAncestor(consoleHandle, GetAncestorFlags.GetRootOwner);
```
Do we need NULL checks here?

mjcheetham replied:
`GetConsoleWindow` will [return `NULL`](https://learn.microsoft.com/en-us/windows/console/getconsolewindow#return-value) (`IntPtr.Zero`) if there is no console window available.

The [docs for `GetAncestor`](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getancestor) don't actually say what happens if you pass `NULL` as the `hwnd` argument. 🤔

However, trying it with `NULL` I get `NULL` back again (`IntPtr.Zero`). Using `IntPtr.Zero` with `WithParentActivityOrWindow` in MSAL results in an error since a window is absolutely required (for WAM).

Having no console window is totally possible, so I guess at this point all we can do is grab a handle to the desktop object?

cc: @bgavrilMS

mjcheetham replied:
Added a new commit that, when the broker is enabled and we are unable to find a parent window, shows a 'dummy' top-level progress bar window, and uses that window handle as the parent.

<img width="801" alt="wam_dummy_window" src="https://user-images.githubusercontent.com/5658207/231294327-526aa279-daee-4f35-9336-a7ed13de4bc2.png">

bgavrilMS replied:
Ideally we should get this type of functionality into MSAL. Apps should need to reference Win.Forms and create windows. Will follow up offline.

bgavrilMS replied:
Note that this solution is imperfect, for example if multiple monitors are used.

bgavrilMS replied:
Another way of doing it is to fail yourself, i.e. if you are hosted in a UI application and the UI application does not initialize GCM correctly with its window handle - fail. If you are hosted in a console app, then the above logic should "just work"

mjcheetham replied:
Since this scenario (no console window and not being told the parent window explicitly) is likely the default for most versions of Visual Studio and tools like SourceTree etc, we're going to need to support this out-of-the box.

I think implementing the 'dummy' progress dialog window should suffice.

ldennington reviewed (approved):
Looks great! This allowed for a lot of nice cleanup.

mjcheetham committed:
msauth: use dummy window for WAM if no parent

If we are unable to get a parent window handle; because for example, we
don't have a console, then create a small 'dummy' window using WinForms
that we can use to pass a handle to MSAL.

ldennington commented on src/shared/Core/Authentication/MicrosoftAuthentication.cs:
```diff
@@ -229,20 +247,36 @@ private async Task<AuthenticationResult> GetAccessTokenSilentlyAsync(IPublicClie
+                        _dummyWindow = new DummyWindow();
+                        appBuilder.WithParentActivityOrWindow(_dummyWindow.ShowAndGetHandle);
+#else
+                        Context.Trace.WriteLine($"Using desktop parent window for MSAL authentication dialogs.");
+                        parentHandle = User32.GetDesktopWindow();
+#endif
```
Is this alternative needed? Since we only need to do this if we're using the broker and the broker is only available on Windows, wouldn't we always be on .NET Framework?

I know there has been some talk with Bogdan about the macOS broker, though, so if you've added it in anticipation of that, then that's fine!

mjcheetham replied:
You are correct that in most cases you'd be using .NET Framework on Windows, however if you're using the .NET Tool on Windows then that's using .NET 6... but then again we'd never enable the broker.. so yes.. I'll remove this 😃

ldennington reviewed (approved):
Latest commit looks good, just a few small comments. Overall, though, I think this solution is all right until (hopefully) the issue is handled in MSAL.

mjcheetham committed:
msauth: update MSAuth flow comment to reflect today

Update the large block comment on how we select which types of flows to
use for Microsoft authentication.

mjcheetham commented on src/shared/Core/Authentication/MicrosoftAuthentication.cs:
```diff
@@ -54,106 +62,116 @@ public MicrosoftAuthentication(ICommandContext context)
+                // On other runtimes (e.g., .NET Core) MSAL only supports the system webview flow (launch the user's browser),
+                // and the device-code flows.
+                //
+                //     Note: .NET Core 3 allows using WinForms when run on Windows but MSAL does not yet support this.
```
Updated to drop this reference.

mjcheetham commented on src/shared/Core/Authentication/MicrosoftAuthentication.cs:
```diff
@@ -54,106 +62,116 @@ public MicrosoftAuthentication(ICommandContext context)
+                // in the first instance.
+                //
+                // On .NET Framework MSAL supports the WinForms based 'embedded' webview UI. For Windows + .NET Framework this is the
+                // best and natural experience.
```
We did, before broker (when enabled). Reworded this entire block to be clearer.

mjcheetham commented:
@ldennington I've updated that outdated block comment in _MicrosoftAuthentication.cs`; did you want to have another :eyes: over it? Thanks!

ldennington reviewed (approved):
Comment updates look good - thanks for doing that!
